### PR TITLE
chore: porting fixes for receiving relay messages

### PR DESCRIPTION
**Pull request by gabrielmer**
Porting the fixes from https://github.com/status-im/status-go/pull/6185 to the go bindings

### CODE REVIEW

**Review by richard-ramos — APPROVED** [review]

File: waku/nwaku.go [hunk 1]
```diff
@@ -496,6 +495,8 @@ func newWakuNode(ctx context.Context, config *WakuConfig) (*WakuNode, error) {
 
 	wg.Add(1)
 	n.wakuCtx = C.cGoWakuNew(cJsonConfig, resp)
+	n.MsgChan = make(chan common.Envelope, 100)
```
Comment (Ivansete-status) [inline waku/nwaku.go:498]: Chan chan, chan chan chan :)
```suggestion
	n.MsgChan = make(chan common.Envelope, MaxMessageChanWakuChanChanChan)
```

File: waku/nwaku.go [hunk 2]
```diff
@@ -532,12 +533,70 @@ func New(nwakuCfg *WakuConfig, logger *zap.Logger) (*Waku, error) {
...
+
 //export globalEventCallback
 func globalEventCallback(callerRet C.int, msg *C.char, len C.size_t, userData unsafe.Pointer) {
-	// This is shared among all Golang instances
-	// TODO-nwaku
-	// self := Waku{wakuCtx: userData}
-	// self.MyEventCallback(callerRet, msg, len)
+	if callerRet == C.RET_OK {
+		eventStr := C.GoStringN(msg, C.int(len))
+		node, ok := nodeRegistry[userData]
```
Comment (Ivansete-status) [inline waku/nwaku.go:563]: May add a comment explaining what info `userData` contains in this point?

File: waku/nwaku.go [hunk 3]
```diff
@@ -532,12 +533,70 @@ func New(nwakuCfg *WakuConfig, logger *zap.Logger) (*Waku, error) {
...
+func (n *WakuNode) OnEvent(eventStr string) {
+	jsonEvent := jsonEvent{}
+	err := json.Unmarshal([]byte(eventStr), &jsonEvent)
+	if err != nil {
+		n.logger.Error("could not unmarshal nwaku event string", zap.Error(err))
+		return
+	}
+
+	switch jsonEvent.EventType {
+	case "message":
```
Comment (Ivansete-status) [inline waku/nwaku.go:589]: Shall we give a more meaningful name?
```suggestion
	case "relay-waku-message":
```

**Review by Ivansete-status — APPROVED** [review]
Gorgeous! Thanks for it! :raised_hands:

File: waku/nwaku.go (hunk 3 above)
Comment (gabrielmer) [reply waku/nwaku.go:589]: this is reading the `message` event defined in libwaku here: https://github.com/waku-org/nwaku/blob/ab0c1d4aa0e545c2c74b7e1ea8045823370eee4d/library/events/json_message_event.nim#L86-L87

We can change the name, but we need to change it too in libwaku, in status-go and in every single repo that is using libwaku to receive messages - this will break compatibility everywhere

So I would probably rename the event at a later stage when integrating the bindings everywhere.
Not sure if it's worth breaking all the repos and modifying them now just because of a name change, I think it's better to do it once we are performing integrations across all repos and take advantage that we're already modifying code in all of them

File: waku/nwaku.go (hunk 2 above)
Comment (gabrielmer) [reply waku/nwaku.go:563]: Ok I got confused and was hard to follow, but I think I got it now

This comment says that the ctx is returned by `WakuCallback` (in the `userData` field) to know to which node an event corresponds
https://github.com/waku-org/waku-go-bindings/blob/8e5ed15dbd0397885fa5ab6340c0cd3ce9ef5b1b/waku/nwaku.go#L536-L542

We can see how we send the context twice over here (one in the `ctx` field and the other in the `userData` one)

https://github.com/waku-org/waku-go-bindings/blob/8e5ed15dbd0397885fa5ab6340c0cd3ce9ef5b1b/waku/nwaku.go#L98-L112

So when calling the event callback, libwaku will have the original context in `eventUserData` and will send it back [here](https://github.com/waku-org/nwaku/blob/ab0c1d4aa0e545c2c74b7e1ea8045823370eee4d/library/libwaku.nim#L76-L79), and we'll know to which node the event corresponds

Uff this is actually pretty hacky and hard to follow, but maybe I'm missing something. I get particularly confused by our use of `userData`, which I understand it was also some other hacky solution to working with Rust's closures.

And I see we also use the `userData` field to hand over a pointer with memory allocated for a response, and now in this case to send back the context.

Idk, I get the idea that the `userData` field is used for many unrelated things and it got pretty messy. Everything seems to work but at some point we should revisit the design and use of this field.

@richard-ramos @Ivansete-status did I understand it well?
